### PR TITLE
IBM MQ source and target

### PR DESCRIPTION
**Pull request by tzununbekov**
In this PR:
- IBM MQ Source and Target
- Makefile updates to support custom-built components such as IBM MQ adapters

Not in this PR:
- Tests
- Validation webhooks for new CRDs
- Source's delivery backoff delay is exposed in the CRD but not handled by the adapter

The things that are missing in this PR will be filled in as follow-up issues to speed up the merge process.

### CODE REVIEW

File: pkg/targets/adapter/ibmmqtarget/adapter.go [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+}
+
+// NewAdapter adapter implementation
+func NewAdapter(ctx context.Context, envAcc pkgadapter.EnvConfigAccessor, ceClient cloudevents.Client) pkgadapter.Adapter {
+	env := envAcc.(*TargetEnvAccessor)
+	logger := logging.FromContext(ctx)
+
+	replier, err := targetce.New(env.Component, logger.Named("replier"),
+		targetce.ReplierWithStatefulHeaders(env.BridgeIdentifier),
+		targetce.ReplierWithStaticResponseType("io.triggermesh.ibmmqtarget.response"),
```
Comment (JeffNeff) [inline pkg/targets/adapter/ibmmqtarget/adapter.go:52]: normally we put the event types in the `/pkg/apis/targets/v1alpha1/` folder under the targets lifecycle file. see here -> https://github.com/triggermesh/triggermesh/blob/f879f93c7d40c0d9f5b5bf9018d11bb545fb6192/pkg/apis/targets/v1alpha1/googlecloudfirestore_lifecycle.go#L29

File: config/300-ibmmqsource.yaml [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+        properties:
+          spec:
+            description: Desired state of the event source.
+            type: object
+            properties:
+              connectionName:
+                description: IBM MQ server URI, e.g. ibm-mq.default.svc.cluster.local(1414).
+                type: string
+              channelName:
+                description: The name of IBM MQ Channel to listen events on.
```
Comment (JeffNeff) [inline config/300-ibmmqsource.yaml:57]:
```suggestion
                description: The name of IBM MQ Channel to listen to events on. Please refer to the IBM MQ documentation for more information. https://www.ibm.com/docs/en/ibm-mq/9.1?topic=types-channels
```

File: config/300-ibmmqsource.yaml [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+                type: string
+              queueManager:
+                description: IBM MQ Queue Manager name.
```
Comment (JeffNeff) [inline config/300-ibmmqsource.yaml:60]:
```suggestion
                description: IBM MQ Queue Manager name. Please refer to the IBM MQ documentation for more information. https://www.ibm.com/docs/en/ibm-mq/9.1?topic=types-mq-queue-managers
```

File: config/300-ibmmqsource.yaml [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+                type: string
+              queueName:
+                description: IBM MQ event source Queue name.
```
Comment (JeffNeff) [inline config/300-ibmmqsource.yaml:63]:
```suggestion
                description: IBM MQ event source Queue name. Please refer to the IBM MQ documentation for more information. https://www.ibm.com/docs/en/ibm-mq/9.1?topic=types-queues
```

File: config/300-ibmmqsource.yaml [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+                          key:
+                            type: string
+                        required:
+                        - name
+                        - key
+                  password:
+                    type: object
+                    properties:
+                      value:
+                        description: Plain text password.
```
Comment (JeffNeff) [inline config/300-ibmmqsource.yaml:132]: im sorry if this sounds stupid.. but the plain text password for what? Where do i get this? Or is it user created? What's it being used for?

File: config/300-ibmmqsource.yaml [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+                - required: [ref]
+                - required: [uri]
+              credentials:
+                type: object
+                properties:
+                  username:
+                    type: object
+                    properties:
+                      value:
+                        description: Plain text user ID.
```
Comment (JeffNeff) [inline config/300-ibmmqsource.yaml:115]: Where does this user ID come from? is there some documentation we can link here?

File: config/301-ibmmqtarget.yaml [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+        properties:
+          spec:
+            description: Desired state of the event target.
+            type: object
+            properties:
+              connectionName:
+                description: IBM MQ server URI, e.g. ibm-mq.default.svc.cluster.local(1414).
+                type: string
+              channelName:
+                description: The name of IBM MQ Channel to listen events on.
```
Comment (JeffNeff) [inline config/301-ibmmqtarget.yaml:61]:
```suggestion
                description: The name of IBM MQ Channel to listen to events on. Please refer to the IBM MQ documentation for more information. https://www.ibm.com/docs/en/ibm-mq/9.1?topic=types-channels
```

File: config/301-ibmmqtarget.yaml [hunk 8]
```diff
@@ -0,0 +1,203 @@
...
+                type: string
+              queueManager:
+                description: IBM MQ Queue Manager name.
```
Comment (JeffNeff) [inline config/301-ibmmqtarget.yaml:64]:
```suggestion
                description: IBM MQ Queue Manager name.  Please refer to the IBM MQ documentation for more information. https://www.ibm.com/docs/en/ibm-mq/9.1?topic=types-mq-queue-managers
```

File: config/301-ibmmqtarget.yaml [hunk 9]
```diff
@@ -0,0 +1,203 @@
...
+                type: string
+              queueName:
+                description: IBM MQ event source Queue name.
```
Comment (JeffNeff) [inline config/301-ibmmqtarget.yaml:67]: im pretty shur i know what your asking for... but i think this description could be confusing to a lot of users. 
 Maybe try:
 
 ```suggestion
                description: The IBM MQ Queue name. 
```

File: config/301-ibmmqtarget.yaml [hunk 10]
```diff
@@ -0,0 +1,203 @@
...
+                  never.'
+                properties:
+                  payloadPolicy:
+                    type: string
+                    enum: [always, error, never]
+              discardCloudEventContext:
+                description: Produce a new cloud event based on the response from the SNS API call.
+                type: boolean
+              replyTo:
+                description: IBM MQ's Reply-to message descriptor parameter.
```
Comment (JeffNeff) [inline config/301-ibmmqtarget.yaml:81]: Is there some documentation we can link to here? Im having troubles digging up something

File: config/301-ibmmqtarget.yaml [hunk 11]
```diff
@@ -0,0 +1,203 @@
...
+                type: object
+                properties: 
+                  queueManager:
+                    description: Queue Manager name for Reply-to parameter.
+                    type: string
+                  queueName:
+                    description: Queue name for Reply-to parameter.
+                    type: string
+              credentials:
```
Comment (JeffNeff) [inline config/301-ibmmqtarget.yaml:90]: Could we add a description here?

File: config/samples/ibm-mq/ibmmq-server.yaml [hunk 12]
```diff
@@ -0,0 +1,39 @@
+apiVersion: apps/v1
```
Comment (JeffNeff) [inline config/samples/ibm-mq/ibmmq-server.yaml:1]: can we not make an ibm-mq folder in the samples? and just have it under `Sources` and `Targets` ? like all the rest of them?

File: pkg/apis/sources/v1alpha1/ibmmq_lifecycle.go [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+	"strings"
+
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"knative.dev/pkg/apis"
+	duckv1 "knative.dev/pkg/apis/duck/v1"
+)
+
+// Managed event types
+const (
+	IBMMQSourceEventType = "io.triggermesh.ibmmqsource"
```
Comment (JeffNeff) [inline pkg/apis/sources/v1alpha1/ibmmq_lifecycle.go:30]: Really not a fan of this event type to be honest @tzununbekov . how about? 
```suggestion
	IBMMQSourceEventType = "io.triggermesh.ibm.mq.message"
```

File: config/300-ibmmqsource.yaml [hunk 14]
```diff
@@ -0,0 +1,225 @@
...
+kind: CustomResourceDefinition
+metadata:
+  name: ibmmqsources.sources.triggermesh.io
+  labels:
+    duck.knative.dev/addressable: 'true'
+    triggermesh.io/crd-install: 'true'
+  annotations:
+    registry.knative.dev/eventTypes: |
+      [
+        { "type": "io.triggermesh.ibmmqsource" }
```
Comment (JeffNeff) [inline config/300-ibmmqsource.yaml:25]:
```suggestion
        { "type": "io.triggermesh.ibm.mq.message" }
```

File: pkg/apis/targets/v1alpha1/ibmmq_lifecycle.go [hunk 15]
```diff
@@ -0,0 +1,116 @@
...
+
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"knative.dev/pkg/apis"
+	duckv1 "knative.dev/pkg/apis/duck/v1"
+	servingv1 "knative.dev/serving/pkg/apis/serving/v1"
+)
+
+// Managed event types
+const (
+	IBMMQTargetGenericRequestEventType  = "io.triggermesh.ibmmqtarget.put"
```
Comment (JeffNeff) [inline pkg/apis/targets/v1alpha1/ibmmq_lifecycle.go:30]:
```suggestion
	IBMMQTargetGenericRequestEventType  = "io.triggermesh.ibm.mq.put"
```

File: pkg/apis/targets/v1alpha1/ibmmq_lifecycle.go [hunk 16]
```diff
@@ -0,0 +1,116 @@
...
+	IBMMQTargetGenericResponseEventType = "io.triggermesh.ibmmqtarget.response"
```
Comment (JeffNeff) [inline pkg/apis/targets/v1alpha1/ibmmq_lifecycle.go:31]:
```suggestion
	IBMMQTargetGenericResponseEventType = "io.triggermesh.ibm.mq.response"
```

File: pkg/targets/adapter/ibmmqtarget/adapter.go (hunk 1 above)
Comment (JeffNeff) [reply pkg/targets/adapter/ibmmqtarget/adapter.go:52]: I see now. you actually do have these pre-existing.. can we use those instead?

**Comment by tzununbekov** [comment]
@JeffNeff thanks for the review, I've handled you comments in 73fb91b2575168d5e03bd97c9dbe9bf352d8c378. Please let me know if I missed something.

File: Makefile [hunk 17]
```diff
@@ -39,13 +43,28 @@ GOLINT            ?= golangci-lint run --timeout 5m
...
+					 $(GOMODULE)/cmd/ibmmqsource-adapter \
+					 $(GOMODULE)/cmd/ibmmqtarget-adapter
+
 LDFLAGS_STATIC     = $(LDFLAGS) -extldflags=-static
 
 HAS_GOTESTSUM     := $(shell command -v gotestsum;)
 HAS_GOLANGCI_LINT := $(shell command -v golangci-lint;)
 
+SED				  := sed -i
+ifeq ($(shell sh -c 'uname 2>/dev/null || echo Unknown'),Darwin)
```
Comment (cab105) [inline Makefile:64]: Good catch although IIRC it isn't just Darwin with the wonky sed, but the BSDs.

File: pkg/targets/adapter/ibmmqtarget/mq/config.go [hunk 18]
```diff
@@ -0,0 +1,39 @@
...
+
+package mq
+
+// EnvConnectionConfig is IBM MQ common connection parameters.
+type EnvConnectionConfig struct {
+	QueueManager   string `envconfig:"QUEUE_MANAGER" default:"QM1"`
+	ChannelName    string `envconfig:"CHANNEL_NAME" default:"DEV.APP.SVRCONN"`
+	ConnectionName string `envconfig:"CONNECTION_NAME" default:"localhost(1414)"`
+	User           string `envconfig:"USER" default:"app"`
+	Password       string `envconfig:"PASSWORD" default:"password"`
```
Comment (cab105) [inline pkg/targets/adapter/ibmmqtarget/mq/config.go:25]: How many of the default values do we need for the target?

File: cmd/ibmmqsource-adapter/Dockerfile [hunk 19]
```diff
@@ -0,0 +1,36 @@
...
+
+RUN --mount=type=cache,target=/root/.cache/go-build \
+    GOOS=linux GOARCH=amd64 go build -v -o ibmmqsource-adapter ./cmd/ibmmqsource-adapter 
+
+
+FROM gcr.io/distroless/base:nonroot
+
+WORKDIR /opt/mqm/
+COPY --from=builder /usr/lib/x86_64-linux-gnu/libstdc++.so.6 /usr/lib/x86_64-linux-gnu/
+COPY --from=builder /lib/x86_64-linux-gnu/libgcc_s.so.1 /lib/x86_64-linux-gnu/
```
Comment (cab105) [inline cmd/ibmmqsource-adapter/Dockerfile:32]: This makes me very nervous given the dynamic linking involved and pulling in the gcc/libstdc++ libraries. Is it possible to use another image besides distroless for this case?

**Review by cab105 — COMMENTED** [review]
Nice! Do see one smallish nit and a not so small question.

File: Makefile (hunk 17 above)
Comment (tzununbekov) [reply Makefile]: ✔️
https://github.com/triggermesh/triggermesh/pull/337/commits/8f88d3efaf87ae86ee4c9a56c2ff1ecf22a99260#diff-76ed074a9305c04054cdebb9e9aad2d818052b07091de1f20cad0bbac34ffb52R65-R66

File: pkg/targets/adapter/ibmmqtarget/mq/config.go (hunk 18 above)
Comment (tzununbekov) [reply pkg/targets/adapter/ibmmqtarget/mq/config.go]: Yeah, not the right place for the default values anyway. Removed.

File: cmd/ibmmqsource-adapter/Dockerfile (hunk 19 above)
Comment (tzununbekov) [reply cmd/ibmmqsource-adapter/Dockerfile]: Replaced with `debian:stable-slim`. Initially, I wanted to make these adapter images as close to the ones that we build with `ko` as possible, but now I don't see much sense in it.

**Review by cab105 — APPROVED** [review]
Thank you for handling the initial nits.

**Review by antoineco — COMMENTED** [review]
The Makefile looks so convoluted with these changes that I wonder if we shouldn't move all the logic to scripts instead (Python, Bash, ...). We aren't helping ourselves hacking around Make features to force complex logic in, and future troubleshooting might get complicated.

File: Makefile [hunk 20]
```diff
@@ -39,14 +43,31 @@ GOLINT            ?= golangci-lint run --timeout 5m
...
 
-.PHONY: help all build release vm-images test lint fmt fmt-test images clean install-gotestsum install-golangci-lint deploy undeploy
+SED               := sed -i
+
+OS                := $(shell sh -c 'uname 2>/dev/null || echo Unknown')
+ifeq ($(OS),$(filter $(OS),Darwin FreeBSD NetBSD))
+    SED           += ''
+endif
+
+.PHONY: sss help all build release vm-images test lint fmt fmt-test images clean install-gotestsum install-golangci-lint deploy undeploy
```
[the remark this answers is not in the document]
Comment (tzununbekov) [reply Makefile:70]: leftovers from some local tests probably, removing

**Comment by tzununbekov** [comment]
> The Makefile looks so convoluted with these changes that I wonder if we shouldn't move all the logic to scripts instead (Python, Bash, ...). We aren't helping ourselves hacking around Make features to force complex logic in, and future troubleshooting might get complicated.

I agree that these changes complicate the Makefile, but I'm not sure what solution would fit better here. Can we take this PR as the way to include IBM MQ integration components into the core and then work on the build improvements? That will help to narrow down the scope of the changes.

**Comment by JeffNeff** [comment]
> > The Makefile looks so convoluted with these changes that I wonder if we shouldn't move all the logic to scripts instead (Python, Bash, ...). We aren't helping ourselves hacking around Make features to force complex logic in, and future troubleshooting might get complicated.
> 
> I agree that these changes complicate the Makefile, but I'm not sure what solution would fit better here. Can we take this PR as the way to include IBM MQ integration components into the core and then work on the build improvements? That will help to narrow down the scope of the changes.

Antoine is asking for that as well here -> https://github.com/triggermesh/triggermesh/issues/173

**Comment by antoineco** [comment]
@tzununbekov Yes sure. Just wanted to point out that we'll need to have a battle plan ready for future additions like this, because it may not be the last one. Who knows, we may even end up writing adapters in other languages for whatever reason!

**Pull request merged**